### DISCUSSION TITLE
How do I type a component wrapping `Link` with `typedRoutes: true`?

### DISCUSSION THREAD

**Discussion by Svish** [vercel/next.js#50611] [App Router] [answered]
I want to wrap the `Link` comment from `next/link` to provide some default styling and other stuff, but I'm unable to figure out how to type the `href` prop.

```ts
import type { ReactNode } from 'react'
import NextLink from 'next/link'

interface Props {
  href: ??; // <-- What type to use here?
  children: ReactNode;
}

export default function MyLink({ href, label }: Props) {
  return (
    <NextLink href={href} className="foo bar">
      {children}
    </NextLink>
  )
}
```

Without `typedRoutes: true` I would just use `href: string`, but if I do that the `Link` component will complain about `href` not being compatible with `string`, and I do want to keep it strongly typed to have the same type-checking on my own component, but I don't understand how to get the correctly typed type for `href`. 🤔

**Comment by devjiwonchoi** [comment]
The type of `href` is embedded inside the `next` module.

```
// your imports
import { Url } from 'next/dist/shared/lib/router/router';

// ...

interface Props {
 href: Url;

// ...
```

For an advanced approach, try this.

```
// your imports
import { UrlObject } from 'url';
declare type Url = string | UrlObject;

// ...

interface Props {
 href: Url;

// ...

**Reply by Svish** [reply]
But that is basically the same as `string`, untyped.

When I look at the `Url` type it's `UrlObject | string`. The actual type of `href` for the `Link` component when I look at it is `__next_route_internal_types__.RouteImpl<T> | UrlObject`, where `T` is some `RouteType` generic I don't know what does, and `__next_route_internal_types__` is some internal unavailable thing... 😕

**Reply by devjiwonchoi** [reply]
Could you please provide a reference?

**Reply by Svish** [reply]
![image](https://github.com/vercel/next.js/assets/142162/0ceaa7dc-4197-4934-9423-c37d815e0cbe)

**Reply by devjiwonchoi** [reply]
Perhaps you've already looked through the files, for the `__next_route_internal_types__` types it is stated:

> Internal types used by the Next.js router and Link component. These types are not meant to be used directly.

so it seems like there's no adequate reason to dig further in for `href`'s type.

For the `__next_route_internal_types__.RouteImpl` it appears to be:

```
type __next_route_internal_types__.RouteImpl<T> = __next_route_internal_types__.StaticRoutes |
 `/?${string}` | `/#${string}` | `/admin?${string}` | `/admin#${string}` | `/admin/settings?${string}` |
 `/admin/settings#${string}` | (T extends never ? T : never)
```

which is basically type `string`.

Since you've stated:

> the Link component will complain about href not being compatible with string

Could you please provide the following lint?

**Reply by Svish** [reply]
So this is basically the component that I want to make:
```ts
import type { ReactNode } from 'react'
import clsx from 'clsx'
import NextLink from 'next/link'

interface Props {
  href: string
  className?: string
  children: ReactNode
}

export default function Link({ href, className, children }: Props) {
  return (
    <NextLink
      href={href}
      className={clsx('my common link classes all links should have', className)}>
      {children}
    </NextLink>
  )
}
```

But, since I have `typedRoutes: true` turned on, I'm getting an error on the `href` here:

![image](https://github.com/vercel/next.js/assets/142162/59eded44-eea3-4ac2-9a78-aa906576d4b6)

**Reply by devjiwonchoi** [reply]
And it still gives you the same lint after using my answer?

**Reply by Svish** [reply]
Yup:

![image](https://github.com/vercel/next.js/assets/142162/0315ce22-ee2a-4e8d-88e9-674b2a653b7c)

**Reply by devjiwonchoi** [reply]
Yes, I've checked it on my local and have the same issue.

Since the `typedRoutes` option is experimental, I'm assuming that this feature hasn't been done yet.

I would like to recommend you disable the `typedRoutes` option for now and wait for this issue to be resolved.

If you don't mind can I create an issue for this? I'll bring it up in this discussion after the work is done.

**Reply by Svish** [reply]
Yeah, the reason why I asked it here was to bring up the issue to the Next.js-team, to see if they had thought about this issue or not. And seems they have not (or just not gotten there yet).

I can create the issue myself, thank you though 🙃

**Comment by Svish** [comment, marked as answer]
Solution given me after I created an issue #50644:

Apparently there is a `Route` type exported from `next` that can be used for this:
```ts
import type { ReactNode } from 'react'
import type { Route } from 'next'
import clsx from 'clsx'
import NextLink from 'next/link'

interface Props<R extends string> {
  href: Route<R>
  className?: string
  children: ReactNode
}

export default function Link<R extends string>({ href, className, children }: Props<R>) {
  return (
    <NextLink
      href={href}
      className={clsx('my common link classes all links should have', className)}>
      {children}
    </NextLink>
  )
}
```

**Reply by Svish** [reply]
After realizing the mystical generic is just `R extends string`, I realized this can be made even simpler by using the `LinkProps` type:
```ts
import type { ReactNode } from 'react'
import clsx from 'clsx'

import NextLink, { type LinkProps } from 'next/link'

export default function Link<R extends string>({ className, children, ...props }: LinkProps<R>) {
  return (
    <NextLink
      href={href}
      className={clsx('my common link classes all links should have', className)}
      {...props}
    >
      {children}
    </NextLink>
  )
}
```

**Reply by devjiwonchoi** [reply]
Outstanding!👊

**Comment by tresorama** [comment]
In case you need to define a variable that contains the value of href, and keep the type-check

```ts
type Post = {
  title: string,
  slug: string,
}

const post: Post = {
  title: 'Bangers',
  slug: 'bangers'
};

const postDerivedData = {
   ...post,
   linkUrl: defineNextLinkHref(`/blog/${post.slug}`)
}

<Link href={postDerivedData.linkUrl} />
```

Here is an hack

```ts
import type { LinkProps } from "next/link";

/**
 * Define a string for a `next/link` `href` prop, and type-check it.  
 * Use this when you need to define a variable that will be passed to a `next/link` component in a later step.  
 * 
 * @example
 * type PostItem = {
 *   title: string;
 *   slug: string;
 *   price: number;
 * };
 * 
 * const post: PostItem = { 
 *   title: 'My Post', 
 *   slug: 'my-post', 
 *   price: 100 
 * };
 * 
 * const postDerivedData = {
 *   isCheap: post.price < 100,
 *   linkHref: defineNextLinkHref(`/blog/${post.slug}`), // this will be type-checked
 * }
 * 
 * const MyButton = () => <Link href={postDerivedData.linkHref}>Read more</Link>;
 * 
 */
export const defineNextLinkHref = <Href extends string>(href: LinkProps<Href>['href']) => href;
```

---

 Previously I was using a type with `satisfies`, but doesn't work with dynamic url (`/my/route/${slug}`)

```ts
const postDerivedData = {
   ...post,
   linkUrl: `/blog/${post.slug}` satisfies NextLinkHref,
}
```

```ts
import type Link from "next/link";

export type NextLinkHref = React.ComponentProps<typeof Link>['href'];
```

**Reply by rui-reevo** [reply]
thanks, but if i want to define a type for `postDerivedData `, i find it also not works for dynamic url

```ts
export interface BreadcrumbItem {
  key: string;
  label?: ReactNode | string;
  url?: ReturnType<typeof defineNextLinkHref>;
  isLoading?: boolean;
}

const breadcrumbs={[
  {
    key: 'account-details',
    label: data?.data?.display_name || 'Account Details',
    url: defineNextLinkHref(`/accounts/${data?.data?.id}`), ## error: Type 'UrlObject | RouteImpl<`/accounts/${string}`>' is not assignable to type 'UrlObject | RouteImpl<string> | undefined'. Type '`/accounts/${string}`' is not assignable to type 'UrlObject | RouteImpl<string> | undefined'.ts(2322)
  },
]}
```

any idea to solve this? thanks

**Reply by tresorama** [reply]
*Edit*: Fixed link


[Codesandox](https://codesandbox.io/p/devbox/vsf5lm?file=%2Fapp%2Fposts%2Fpage.tsx%3A28%2C7-28%2C18)
---

You can add any as generic.  
`TestTwo` in sandbox.  

```tsx
export interface BreadcrumbItem {
  key: string;
  label?: ReactNode | string;
  url?: ReturnType<typeof defineNextLinkHref<any>>; // added "any" as generic
  isLoading?: boolean;
}
```

---

Or you can infer `BreadcrumbItem` from a plain ts object.  
`TestThree` in the sandbox.  

```tsx
const breadcrumbs = [
  {
    key: 'account-details',
    label: data?.data?.display_name || 'Account Details',
    url: defineNextLinkHref(`/accounts/${data?.data?.id}`), 
  },
] as const;

export type BreadcrumbItem = typeof breadcrumbs[number];
```
